File: automedal/agent/providers/openai.py
"""OpenAI provider — wraps the official `openai` SDK.

Used for OpenAI direct, Ollama (`/v1`), OpenRouter, Groq, and any other
OpenAI-compatible endpoint via `base_url`.

Translates between our internal Anthropic-flavored message shape and
OpenAI's chat-completions schema:

    internal "tool" role  → OpenAI "tool" role with tool_call_id
    internal assistant tool_use blocks → assistant message with `tool_calls`
"""

from __future__ import annotations

import json
import uuid
from dataclasses import dataclass, field
from typing import Any

from automedal.agent.providers.base import ChatProvider, ChatTurn, ToolCall, Usage
from automedal.agent.retry import with_retry


@dataclass
class OpenAIProvider:
    model: str
    api_key: str
    base_url: str | None = None
    timeout: int = 120

    def _client(self):
        from openai import AsyncOpenAI
        kw: dict[str, Any] = {"api_key": self.api_key, "timeout": self.timeout}
        if self.base_url:
            kw["base_url"] = self.base_url
        return AsyncOpenAI(**kw)

    # ── outbound serialization ───────────────────────────────────────────────

    @staticmethod
    def _to_openai_messages(system: str, internal: list[dict]) -> list[dict]:
        out: list[dict] = []
        if system:
            out.append({"role": "system", "content": system})

        for msg in internal:
            role = msg.get("role")

            if role == "tool":
                out.append({
                    "role": "tool",
                    "tool_call_id": msg["tool_use_id"],
                    "content": msg.get("content", ""),
                })
                continue

            if role == "user":
                c = msg["content"]
                if isinstance(c, str):
                    out.append({"role": "user", "content": c})
                else:
                    # Anthropic-shape tool_result blocks coming through
                    # the internal channel — flatten each as a tool message.
                    flushed = False
                    for block in c:
                        if isinstance(block, dict) and block.get("type") == "tool_result":
                            out.append({
                                "role": "tool",
                                "tool_call_id": block["tool_use_id"],
                                "content": block.get("content", "") if isinstance(block.get("content"), str)
                                           else json.dumps(block.get("content", "")),
                            })
                            flushed = True
                    if not flushed:
                        # User content given as raw block list (rare). Concatenate text.
                        text = "".join(b.get("text", "") for b in c if isinstance(b, dict) and b.get("type") == "text")
                        out.append({"role": "user", "content": text})
                continue

            if role == "assistant":
                blocks = msg.get("content", [])
                text_parts: list[str] = []
                tool_calls: list[dict] = []
                for b in blocks:
                    btype = b.get("type")
                    if btype == "text":
                        text_parts.append(b.get("text", ""))
                    elif btype == "tool_use":
                        tool_calls.append({
                            "id": b["id"],
                            "type": "function",
                            "function": {
                                "name": b["name"],
                                "arguments": json.dumps(b.get("input", {})),
                            },
                        })
                    # ignore thinking / redacted_thinking — OpenAI has no analog
                m: dict[str, Any] = {"role": "assistant"}
                if text_parts:
                    m["content"] = "".join(text_parts)
                if tool_calls:
                    m["tool_calls"] = tool_calls
                if "content" not in m:
                    m["content"] = None
                out.append(m)
                continue

            raise ValueError(f"unknown message role: {role!r}")

        return out

    @staticmethod
    def _tool_specs(tools: list) -> list[dict]:
        return [
            {"type": "function", "function": {
                "name": t.name,
                "description": t.description,
                "parameters": t.schema,
            }} for t in tools
        ]

    # ── chat ─────────────────────────────────────────────────────────────────

    async def chat_stream(
        self,
        *,
        system: str,
        messages: list[dict],
        tools: list,
        events: Any,
    ) -> ChatTurn:
        client = self._client()
        api_messages = self._to_openai_messages(system, messages)

        kwargs: dict[str, Any] = {
            "model": self.model,
            "messages": api_messages,
            "stream": True,
            "stream_options": {"include_usage": True},
        }
        if tools:
            kwargs["tools"] = self._tool_specs(tools)

        text_acc: list[str] = []
        # tool_calls accumulator — OpenAI streams them as deltas keyed by index
        tc_acc: dict[int, dict] = {}
        usage = Usage()
        stop_reason = ""

        try:
            stream = await with_retry(
                lambda: client.chat.completions.create(**kwargs),
                label=f"openai.chat_stream model={self.model}",
                events=events,
            )
            async for chunk in stream:
                # Usage may arrive on its own with empty choices
                if getattr(chunk, "usage", None):
                    u = chunk.usage
                    usage.in_tokens = getattr(u, "prompt_tokens", 0) or 0
                    usage.out_tokens = getattr(u, "completion_tokens", 0) or 0

                if not chunk.choices:
                    continue
                ch0 = chunk.choices[0]

                if getattr(ch0, "finish_reason", None):
                    stop_reason = ch0.finish_reason

                delta = getattr(ch0, "delta", None)
                if delta is None:
                    continue

                txt = getattr(delta, "content", None)
                if txt:
                    text_acc.append(txt)
                    if events is not None:
                        events.delta(txt)

                tcs = getattr(delta, "tool_calls", None) or []
                for tc_delta in tcs:
                    idx = getattr(tc_delta, "index", 0) or 0
                    slot = tc_acc.setdefault(idx, {"id": "", "name": "", "args_str": ""})
                    if getattr(tc_delta, "id", None):
                        slot["id"] = tc_delta.id
                    fn = getattr(tc_delta, "function", None)
                    if fn is not None:
                        if getattr(fn, "name", None):
                            slot["name"] += fn.name
                        if getattr(fn, "arguments", None):
                            slot["args_str"] += fn.arguments
        finally:
            try:
                await client.close()
            except Exception:
                pass

        # Build assistant_blocks + tool_calls
        assistant_blocks: list[dict] = []
        text = "".join(text_acc)
        if text:
            assistant_blocks.append({"type": "text", "text": text})

        tool_calls: list[ToolCall] = []
        for idx in sorted(tc_acc):
            slot = tc_acc[idx]
            args: dict = {}
            try:
                args = json.loads(slot["args_str"]) if slot["args_str"] else {}
            except json.JSONDecodeError:
                args = {"_raw": slot["args_str"]}
            tc_id = slot["id"] or f"call_{uuid.uuid4().hex[:12]}"
            tool_calls.append(ToolCall(id=tc_id, name=slot["name"], args=args))
            assistant_blocks.append({
                "type": "tool_use", "id": tc_id, "name": slot["name"], "input": args,
            })

        return ChatTurn(
            assistant_blocks=assistant_blocks,
            assistant_text=text,
            tool_calls=tool_calls,
            usage=usage,
            stop_reason=stop_reason,
        )

File: internal/agent/providers/openai.go
// OpenAI provider — wraps github.com/openai/openai-go.
//
// Used for OpenAI direct, Ollama (/v1), OpenRouter, Groq, and any other
// OpenAI-compatible endpoint via BaseURL.
//
// Translates between our internal Anthropic-flavoured message shape and
// OpenAI's chat-completions schema:
//
//	internal "tool" role  → OpenAI "tool" role with tool_call_id
//	internal assistant tool_use blocks → assistant message with tool_calls
package providers

import (
	"context"
	"encoding/json"
	"fmt"
	"time"

	"github.com/Flameingmoy/automedal/internal/agent"
	openai "github.com/openai/openai-go"
	openaiopt "github.com/openai/openai-go/option"
	"github.com/openai/openai-go/shared"
)

// OpenAIProvider talks to api.openai.com or any OpenAI-compatible
// endpoint via BaseURL.
type OpenAIProvider struct {
	ModelName string
	APIKey    string
	BaseURL   string
	Timeout   time.Duration
}

func (p *OpenAIProvider) Model() string { return p.ModelName }

func (p *OpenAIProvider) client() openai.Client {
	opts := []openaiopt.RequestOption{openaiopt.WithAPIKey(p.APIKey)}
	if p.BaseURL != "" {
		opts = append(opts, openaiopt.WithBaseURL(p.BaseURL))
	}
	if p.Timeout > 0 {
		opts = append(opts, openaiopt.WithRequestTimeout(p.Timeout))
	}
	return openai.NewClient(opts...)
}

// ChatStream sends one streaming turn through the chat-completions API.
func (p *OpenAIProvider) ChatStream(ctx context.Context, in ChatRequest) (*ChatTurn, error) {
	c := p.client()
	apiMsgs, err := toOpenAIMessages(in.System, in.Messages)
	if err != nil {
		return nil, err
	}
	params := openai.ChatCompletionNewParams{
		Model:    openai.ChatModel(p.ModelName),
		Messages: apiMsgs,
		StreamOptions: openai.ChatCompletionStreamOptionsParam{
			IncludeUsage: openai.Bool(true),
		},
	}
	if len(in.Tools) > 0 {
		params.Tools = toOpenAITools(in.Tools)
	}

	final, err := agent.WithRetry(ctx, func() (openai.ChatCompletion, error) {
		return p.runStream(ctx, &c, params, in.Events)
	}, agent.RetryOpts{
		Sink:  in.Events,
		Label: "openai.chat_stream model=" + p.ModelName,
	})
	if err != nil {
		return nil, err
	}
	return openAIFinalToTurn(&final), nil
}

func (p *OpenAIProvider) runStream(
	ctx context.Context,
	c *openai.Client,
	params openai.ChatCompletionNewParams,
	events *agent.EventSink,
) (openai.ChatCompletion, error) {
	stream := c.Chat.Completions.NewStreaming(ctx, params)
	defer stream.Close()

	acc := openai.ChatCompletionAccumulator{}
	for stream.Next() {
		chunk := stream.Current()
		if !acc.AddChunk(chunk) {
			return openai.ChatCompletion{}, fmt.Errorf("openai accumulator rejected chunk")
		}
		// Live text deltas to the EventSink.
		if events != nil && len(chunk.Choices) > 0 {
			if d := chunk.Choices[0].Delta.Content; d != "" {
				events.Delta(d)
			}
		}
	}
	if err := stream.Err(); err != nil {
		return openai.ChatCompletion{}, err
	}
	return acc.ChatCompletion, nil
}

// ── outbound message translation ────────────────────────────────────────

func toOpenAIMessages(system string, internal []agent.Message) ([]openai.ChatCompletionMessageParamUnion, error) {
	var out []openai.ChatCompletionMessageParamUnion
	if system != "" {
		out = append(out, openai.SystemMessage(system))
	}

	for _, m := range internal {
		switch m.Role {
		case "tool":
			content, _ := m.Content.(string)
			out = append(out, openai.ToolMessage(content, m.ToolUseID))

		case "user":
			switch c := m.Content.(type) {
			case string:
				out = append(out, openai.UserMessage(c))
			case []agent.Block:
				// Anthropic-shape tool_result blocks → flatten as tool messages.
				flushed := false
				for _, b := range c {
					if b.Type == "tool_result" {
						s, _ := b.Content.(string)
						out = append(out, openai.ToolMessage(s, b.ToolUseID))
						flushed = true
					}
				}
				if !flushed {
					var text string
					for _, b := range c {
						if b.Type == "text" {
							text += b.Text
						}
					}
					out = append(out, openai.UserMessage(text))
				}
			default:
				return nil, fmt.Errorf("unexpected user content type %T", m.Content)
			}

		case "assistant":
			blocks, _ := m.Content.([]agent.Block)
			ap := openai.ChatCompletionAssistantMessageParam{}
			var text string
			for _, b := range blocks {
				switch b.Type {
				case "text":
					text += b.Text
				case "tool_use":
					argsJSON, err := json.Marshal(b.Input)
					if err != nil {
						return nil, err
					}
					ap.ToolCalls = append(ap.ToolCalls, openai.ChatCompletionMessageToolCallParam{
						ID: b.ID,
						Function: openai.ChatCompletionMessageToolCallFunctionParam{
							Name:      b.Name,
							Arguments: string(argsJSON),
						},
					})
					// thinking / redacted_thinking: no OpenAI analog; drop.
				}
			}
			if text != "" {
				ap.Content.OfString = openai.String(text)
			}
			out = append(out, openai.ChatCompletionMessageParamUnion{OfAssistant: &ap})

		default:
			return nil, fmt.Errorf("unknown role %q", m.Role)
		}
	}
	return out, nil
}

func toOpenAITools(specs []ToolSpec) []openai.ChatCompletionToolParam {
	out := make([]openai.ChatCompletionToolParam, 0, len(specs))
	for _, s := range specs {
		fd := shared.FunctionDefinitionParam{
			Name:       s.Name,
			Parameters: shared.FunctionParameters(s.Schema),
		}
		if s.Description != "" {
			fd.Description = openai.String(s.Description)
		}
		out = append(out, openai.ChatCompletionToolParam{Function: fd})
	}
	return out
}

// ── inbound (final → ChatTurn) ──────────────────────────────────────────

func openAIFinalToTurn(cc *openai.ChatCompletion) *ChatTurn {
	turn := &ChatTurn{
		Usage: agent.Usage{
			InTokens:  int(cc.Usage.PromptTokens),
			OutTokens: int(cc.Usage.CompletionTokens),
		},
	}
	if len(cc.Choices) == 0 {
		return turn
	}
	choice := cc.Choices[0]
	turn.StopReason = string(choice.FinishReason)

	if text := choice.Message.Content; text != "" {
		turn.AssistantBlocks = append(turn.AssistantBlocks, agent.Block{Type: "text", Text: text})
		turn.AssistantText = text
	}
	for _, tc := range choice.Message.ToolCalls {
		var args map[string]any
		if tc.Function.Arguments != "" {
			if err := json.Unmarshal([]byte(tc.Function.Arguments), &args); err != nil {
				args = map[string]any{"_raw": tc.Function.Arguments}
			}
		}
		turn.ToolCalls = append(turn.ToolCalls, ToolCall{
			ID:   tc.ID,
			Name: tc.Function.Name,
			Args: args,
		})
		turn.AssistantBlocks = append(turn.AssistantBlocks, agent.Block{
			Type:  "tool_use",
			ID:    tc.ID,
			Name:  tc.Function.Name,
			Input: args,
		})
	}
	return turn
}
